src/graph/graphcontext.c

```diff
@@ -45,33 +37,76 @@ GraphContext *GraphContext_New(RedisModuleCtx *ctx, const char *graphname,
 
 	gc->string_mapping = array_new(char *, 64);
 	gc->attributes = raxNew();
-
+	Graph_SetMatrixPolicy(gc->g, SYNC_AND_MINIMIZE_SPACE);
 	QueryCtx_SetGraphCtx(gc);
 
-	// Set and close GraphContext key in Redis keyspace
-	RedisModule_ModuleTypeSetValue(key, GraphContextRedisModuleType, gc);
+	return gc;
+}
+
+/* This method tries to get a graph context, and if it does not exists, create a new one.
+ * The try-get-create flow is done when module global lock is acquired, to enforce consistency
+ * while BGSave is called. */
+GraphContext *_GraphContext_GetOrCreate(RedisModuleCtx *ctx, const char *graphname,
+										int readWriteFlag, size_t node_cap, size_t edge_cap) {
+	GraphContext *gc = NULL;
+	RedisModuleString *rs_name = RedisModule_CreateString(ctx, graphname, strlen(graphname));
+
+	// Acquire lock.
+	assert(pthread_mutex_lock(&_module_mutex) == 0);
```

Comment:
Try to minimising the amount of work done once this lock is acquired,
once you know the key doesn't exists, create GraphContext and only then acquire the lock, worst case you would free the graph context incase the key exists after entering the critical section.

At the moment we're acquiring the lock prior to first retrieval, try opening the key before locking.

Reply:
done

Reply:
I don't know if I agree with this. If we're coming from a QUERY or PROFILE, we're holding the GIL currently anyway. I think we should standardize all the endpoints with one lock or another, since now all of them can modify the keyspace in this region.